File: web-client/src/core/Polynomial5.ts
export interface CamProfile {
  s: number; // Displacement
  v: number; // Velocity
  a: number; // Acceleration
  j: number; // Jerk
  theta: number; // Master axis position
}

export class Polynomial5 {
  masterStart: number;
  masterEnd: number;
  slaveStart: number;
  slaveEnd: number;

  constructor(
    masterStart: number,
    masterEnd: number,
    slaveStart: number,
    slaveEnd: number
  ) {
    this.masterStart = masterStart;
    this.masterEnd = masterEnd;
    this.slaveStart = slaveStart;
    this.slaveEnd = slaveEnd;
  }

  // Calculate the profile at a specific master position theta
  calculate(theta: number): CamProfile {
    // 1. Normalize theta to tau [0, 1]
    const beta = this.masterEnd - this.masterStart;
    
    // Handle out of bounds (clamping) or return 0 for R-R outside range
    if (theta < this.masterStart) return { s: this.slaveStart, v: 0, a: 0, j: 0, theta };
    if (theta > this.masterEnd) return { s: this.slaveEnd, v: 0, a: 0, j: 0, theta };

    const tau = (theta - this.masterStart) / beta;
    const h = this.slaveEnd - this.slaveStart;

    // 2. 5th Order Polynomial Coefficients (Rest-in-Rest)
    // s(tau) = 10*tau^3 - 15*tau^4 + 6*tau^5
    
    // Displacement
    const s_norm = 10 * Math.pow(tau, 3) - 15 * Math.pow(tau, 4) + 6 * Math.pow(tau, 5);
    const s = this.slaveStart + h * s_norm;

    // Velocity (First Derivative)
    // v_norm = 30*tau^2 - 60*tau^3 + 30*tau^4
    // v = (h / beta) * v_norm
    const v_norm = 30 * Math.pow(tau, 2) - 60 * Math.pow(tau, 3) + 30 * Math.pow(tau, 4);
    const v = (h / beta) * v_norm;

    // Acceleration (Second Derivative)
    // a_norm = 60*tau - 180*tau^2 + 120*tau^3
    // a = (h / beta^2) * a_norm
    const a_norm = 60 * tau - 180 * Math.pow(tau, 2) + 120 * Math.pow(tau, 3);
    const a = (h / Math.pow(beta, 2)) * a_norm;

    // Jerk (Third Derivative)
    // j_norm = 60 - 360*tau + 360*tau^2
    // j = (h / beta^3) * j_norm
    const j_norm = 60 - 360 * tau + 360 * Math.pow(tau, 2);
    const j = (h / Math.pow(beta, 3)) * j_norm;

    return { s, v, a, j, theta };
  }

  // Generate discrete points for the entire range
  generateTable(resolution: number = 100): CamProfile[] {
    const data: CamProfile[] = [];
    const step = (this.masterEnd - this.masterStart) / (resolution - 1);

    for (let i = 0; i < resolution; i++) {
        const theta = this.masterStart + i * step;
        data.push(this.calculate(theta));
    }
    return data;
  }
}

File: quintic/core/polynomial5.py
import math


class Polynomial5:
    """
    Implements a 5th-order polynomial motion law (Rest-in-Rest) as per VDI 2143.
    This provides a smooth transition with zero velocity and acceleration at the start and end.

    Formula: s(tau) = 10*tau^3 - 15*tau^4 + 6*tau^5
    where tau is the normalized time/position (0 to 1).
    """

    def __init__(
        self,
        master_start: float,
        master_end: float,
        slave_start: float,
        slave_end: float,
    ):
        self.master_start = master_start
        self.master_end = master_end
        self.slave_start = slave_start
        self.slave_end = slave_end

    def calculate(self, theta: float):
        """
        Calculate displacement (s), velocity (v), acceleration (a), and jerk (j)
        at a specific master position theta.
        """
        # 1. Normalize theta to tau [0, 1]
        beta = self.master_end - self.master_start

        # Avoid division by zero
        if beta == 0:
            return 0.0, 0.0, 0.0, 0.0

        # Handle out of bounds (clamping) for R-R
        if theta < self.master_start:
            return self.slave_start, 0.0, 0.0, 0.0
        if theta > self.master_end:
            return self.slave_end, 0.0, 0.0, 0.0

        tau = (theta - self.master_start) / beta
        h = self.slave_end - self.slave_start

        # 2. 5th Order Polynomial Coefficients (Rest-in-Rest)
        # s_norm(tau) = 10*tau^3 - 15*tau^4 + 6*tau^5

        # Displacement (s)
        # s = start + h * s_norm
        s_norm = 10 * (tau**3) - 15 * (tau**4) + 6 * (tau**5)
        s = self.slave_start + h * s_norm

        # Velocity (v) - First Derivative
        # v_norm = 30*tau^2 - 60*tau^3 + 30*tau^4
        # v = (h / beta) * v_norm
        v_norm = 30 * (tau**2) - 60 * (tau**3) + 30 * (tau**4)
        v = (h / beta) * v_norm

        # Acceleration (a) - Second Derivative
        # a_norm = 60*tau - 180*tau^2 + 120*tau^3
        # a = (h / beta^2) * a_norm
        a_norm = 60 * tau - 180 * (tau**2) + 120 * (tau**3)
        a = (h / (beta**2)) * a_norm

        # Jerk (j) - Third Derivative
        # j_norm = 60 - 360*tau + 360*tau^2
        # j = (h / beta^3) * j_norm
        j_norm = 60 - 360 * tau + 360 * (tau**2)
        j = (h / (beta**3)) * j_norm

        return s, v, a, j

    def generate_table(self, resolution: int = 100):
        """
        Generate a table of discrete points for the entire range.
        """
        data = []
        if resolution < 2:
            return data

        step = (self.master_end - self.master_start) / (resolution - 1)

        for i in range(resolution):
            theta = self.master_start + i * step
            s, v, a, j = self.calculate(theta)
            data.append({"theta": theta, "s": s, "v": v, "a": a, "j": j})
        return data
